_posts/2017-02-19-simplified-database-migrations-in-go.md

```diff
@@ -0,0 +1,98 @@
+---
+layout: post
+title: "Simplified database migrations in go"
+ghavatar: 501424
+ghname: mastoj
+jobtitle: Platform Engineer
+tags: [go, golang]
+---
+
+## Introduction
+
+There are two main criteria that one should look for when choosing a database migration framework; it should be simple and it should always roll forward. When we began looking for a migration framework to use in one of our applications we couldn't find one that was simple in the sense that it could read from file as well as from pre-compiled assets, which we tend to use at Unacast, and only rolls forward. As a result of that we ended up implementing our own simple migration framework for go applications, https://github.com/unacast/migrations, which I will guide you through in this post.
+
+*TLDR; Grab the code here: https://github.com/unacast/migrations*
+
+## Why you shouldn't roll back a database
+
+For some it might sound a little bit odd to say that you should never rollback a database, but let's think about what a rollback is for a moment. While you think about I'll define what I mean rollback and rollforward are.
+
+* Rollforward: an action in which you take your database from one state to another
+* Rollback: an action in which you take your database form one state to another
+
+The observant reader will notice that the definition for both actions are quite similar, or wait, they are the same. This basically means that instead of having a rollback you can just write a new rollforward if you are in a situation where a database update has gone wrong. 
+
+Another reason is that database roll backs are complicated and there are a lot of things that you need to consider, which you most likely won't, when you're writing a rollback. Let me show an example of something that most people would do when writing a new migration adding a new column:
+
+1. Write migrate up that adds column X
+2. Write migrate down that removes column X
+3. Runs migration
+
+So far so good. If step 3 fails you can just rollback right away, no stress, but this could also be solved by using a transaction when you run the migration. A more complicated scenario, and probably more realistic one as well, is that migration went well but a couple of hours later the database starts to act weird. What do you do? Should you run the rollback? Probably not. At this point in time you have already gotten some data in the new column X which you most likely want to keep, so the solution for rolling back isn't just to remove what you just added. Instead you should write a new migration where you define what you should do to not lose any data.
+
+Not everyone agrees with always rolling forward, and that is ok, but this is the way we at Unacast think of migrations.
```

Comment:
I agree, but I think I'll keep application up and downs out of this post. It is not the same issues with applications since they are most of the time stateless, making them easier to go both up and down.